plutus-tx/src/Language/PlutusTx/Lift.hs

```diff
@@ -24,36 +27,47 @@ import qualified Language.PlutusIR.MkPir                       as PIR
 
 import qualified Language.PlutusCore                           as PLC
 import qualified Language.PlutusCore.Constant.Dynamic          as PLC
+import           Language.PlutusCore.Pretty                    (Pretty)
 import           Language.PlutusCore.Quote
 import qualified Language.PlutusCore.StdLib.Data.Function      as PLC
 import qualified Language.PlutusCore.StdLib.Meta.Data.Function as PLC
 
+import           Codec.Serialise
 import           Control.Exception
 import           Control.Monad.Except                          hiding (lift)
 import           Control.Monad.Reader                          hiding (lift)
 
 import           Data.Functor                                  (void)
 import           Data.Proxy
+import qualified Data.Typeable                                 as GHC
+
+type Throwable uni = (PLC.GShow uni, PLC.Closed uni, uni `PLC.Everywhere` Pretty, GHC.Typeable uni)
 
 -- | Get a Plutus Core term corresponding to the given value.
-safeLift :: (Lift.Lift a, AsError e (Provenance ()), MonadError e m, MonadQuote m) => a -> m (PLC.Term TyName Name ())
+safeLift
+    :: (Lift.Lift uni a, AsError e uni (Provenance ()), MonadError e m, MonadQuote m)
+    => a -> m (PLC.Term TyName Name uni ())
 safeLift x = do
     lifted <- liftQuote $ runDefT () $ Lift.lift x
     compiled <- flip runReaderT defaultCompilationCtx $ compileTerm lifted
     pure $ void compiled
 
 -- | Get a Plutus Core program corresponding to the given value.
-safeLiftProgram :: (Lift.Lift a, AsError e (Provenance ()), MonadError e m, MonadQuote m) => a -> m (PLC.Program TyName Name ())
+safeLiftProgram
+    :: (Lift.Lift uni a, AsError e uni (Provenance ()), MonadError e m, MonadQuote m)
+    => a -> m (PLC.Program TyName Name uni ())
 safeLiftProgram x = PLC.Program () (PLC.defaultVersion ()) <$> safeLift x
 
-safeLiftCode :: (Lift.Lift a, AsError e (Provenance ()), MonadError e m, MonadQuote m) => a -> m (CompiledCode a)
+safeLiftCode :: (Lift.Lift uni a, AsError e uni (Provenance ()), MonadError e m, MonadQuote m) => a -> m (CompiledCode uni a)
 safeLiftCode x = DeserializedCode <$> safeLiftProgram x <*> pure Nothing
 
 safeConstCode
-    :: (Lift.Typeable a, AsError e (Provenance ()), MonadError e m, MonadQuote m)
+    :: ( Lift.Typeable uni a, AsError e uni (Provenance ()), MonadError e m, MonadQuote m
+       , PLC.Closed uni, uni `PLC.Everywhere` Serialise
```

Comment:
That `Closed` is there just to circumvent the annoying behaviour of GHC, see this [comment](https://github.com/input-output-hk/plutus/issues/581#issuecomment-582483376). Basically, we can just define `Everywhere` as a class instead of an associated type family and then we don't need that `Closed` and other things become nicer as well (and some [worse](https://github.com/input-output-hk/plutus/issues/581#issuecomment-581884602)), but then things like `deriving Eq/Show/Whatever` break and we have to use standalone deriving declarations everywhere, which is super annoying.

It would be great if we could just get rid of `Closed`.

Reply:
Here's a [branch without `Closed`](https://github.com/effectfully/plutus-prototype/tree/effectfully/builtins/add-extensible-builtins-no-closed) (`Universe.Core` and `Universe.Default` type check, PLC AST stuff doesn't, 'cause non-standalone `deriving` is broken).

@reactormonk you were interested ^